### DISCUSSION TITLE
Corrupted file after download

### DISCUSSION THREAD

**Discussion by GarikFirst** [micropython/micropython#13704] [General]
After stream downloading and checking hash of file I can see that hash is changed, but I don’t understand why. Here is code snippet

import hashlib
import requests

```
# Checking hash of file that web server serve
with open("original/firmware.tar.gz", "rb") as file:
    filehash1 = hashlib.sha1()

    while True:
        chunk = file.read(1024)
        if not chunk:
            break
        filehash1.update(chunk)

    print("original file hash:", filehash1.digest())


# Download and save in chunk this file
resp = requests.get("http://192.168.0.183:8080/test.tar.gz")

stream = resp.raw

with open("firmware.tar.gz", "wb") as fd:
    while True:
        chunk = stream.read(1024)
        if not chunk:
            break
        fd.write(chunk)

resp.close()

# Checking of hash of what we downloaded
with open("firmware.tar.gz", "rb") as fr:
    filehash2 = hashlib.sha1()

    while True:
        chunk = fr.read(1024)
        if not chunk:
            break
        filehash2.update(chunk)

    print("downloade file hash:", filehash2.digest())
```

Output is:
```
original file hash: b'\xa0\xc9e\x92\xa0\x9dq`/\xfa\x92\x01$\x08\xb3\xeb\x85G\x94\xfb'
downloade file hash: b'&\x16\xe0\xe6\x95H~\x0fO\xc5\xdd\xc1\x8cW\xdc\xf9\\\xe5\x85\x1a'
```

What am I doing wrong?

**Comment by GitHubsSilverBullet** [comment]
> What am I doing wrong?

Is your file divisible by 1024? Without remainder?

**Reply by GarikFirst** [reply]
Yeah, I get it… Of course it is not… but I have tried it with read(1) - and no luck, so I assume this is not the point

**Reply by GitHubsSilverBullet** [reply]
Well, I would reduce the stuff to the bare minimum.
Create a testfile, 256bytes in size, containing all values 0…255 and use that as a test case.
You can download it completely and inspect it for any inconsistencies. (specific values dropped or changed)
I'd look for 0x00, 0x0a, 0x0d or more broadly all values in the 0x00…0x1f range. Also everything above the ASCII range (0x7e and up)

**Reply by GarikFirst** [reply]
Thank you, I’ll try to do deep inspection of content

**Comment by GarikFirst** [comment]
Somehow it starts working after rewriting from scratch, so I think I can close this now
